picard/file.py

```diff
@@ -323,6 +325,38 @@ def _fixed_splitext(self, filename):
             new_filename = ''
         return new_filename, ext
 
+    def _format_filename(self, new_dirname, new_filename, metadata, settings):
+        # TODO: tests !!
+        new_filename, ext = self._fixed_splitext(new_filename)
+        ext = ext.lower()
+        new_filename = new_filename + ext
+
+        # expand the naming format
+        naming_format = settings['file_naming_format']
+        if naming_format:
+            new_filename = self._script_to_filename(naming_format, metadata, settings)
+            # NOTE: the _script_to_filename strips the extension away
+            new_filename = new_filename + ext
+            if not settings['move_files']:
+                new_filename = os.path.basename(new_filename)
+            win_compat = IS_WIN or settings['windows_compatibility']
+            new_filename = make_short_filename(new_dirname, new_filename,
+                                               win_compat)
+            # TODO: move following logic under util.filenaming
+            # (and reconsider its necessity)
+            # win32 compatibility fixes
+            if win_compat:
+                new_filename = new_filename.replace('./', '_/').replace('.\\', '_\\')
```

Comment:
```
>>> filename ='a./b./c.'
>>> filename.replace('./', '_/').replace('.\\', '_\\')
'a_/b_/c.'
```

The last dot isn't replaced.

Reply:
Yes, this should be considered if we move it to a separate function. But in the current code this is a non-issue, because the extension is always added here.

But I gave this some testing on Windows: Windows actually just treats a file ending with a dot as if it has no dot at the end (I think Windows understands this as a file with empty extension, so `test` and `test.` are actually the same). If you try to create something like `C:\test.\afile.` it will end up as `C:\test\afile`. Actually you can access a file with as many dots as you want at the end, i doesn't matter if you do `open("somefile")` or `open("somefile......")`. 

Instead of replacing the dot at the end with an `_` I think we could also just remove all trailing dots from each path component. But one use case definitely also was to somewhat preserve three dots at the end of e.g. release or song titles. Not sure whether I actually prefer "The Title.._" over just "The Title", though.

Reply:
> Not sure whether I actually prefer "The Title.._" over just "The Title", though.

Since, on Windows, path lengths are rather limited, i'd go for removing all trailing dots rather than replacing them.

Reply:
Agreed, yes. It's also in line with what Windows would do if you tried to name a file like that, so this should be easy to communicate to users.

Reply:
I'd rather merge this (refactoring) PR, before changing actual code, it'd be simple to review